### ISSUE TITLE
Is there a way perform inference on models with custom trainer classes that doesn't require modifying `nnunetv2` package files?

### ISSUE THREAD

**Issue by joshuacwnewton** [MIC-DKFZ/nnUNet#2856]
The documentation states:

https://github.com/MIC-DKFZ/nnUNet/blob/f1851fbaf2c53dcb51b079b60a01de528a7d0c17/documentation/extending_nnunet.md?plain=1#L15-L19

And the relevant inference code is:

https://github.com/MIC-DKFZ/nnUNet/blob/f1851fbaf2c53dcb51b079b60a01de528a7d0c17/nnunetv2/inference/predict_from_raw_data.py#L97-L111

However, by hardcoding the directory `nunetv2.__path__[0]`, it would seem that the new trainer class must be put inside of the `nnunetv2` package folder prior to runtime.

This is causing us to consider a lot of hacky workarounds in order to make inference work for models trained using custom trainers:

- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4955

> If the trainer class isn't inside the installed nnunetv2 library, then the inference will crash.

Just as a sanity check -- is our understanding accurate? Is it necessary to modify `nnunetv2` package files? Or is there a way to pass a trainer class from our own code (to prevent the crash)?

**Comment by jcohenadad** [comment]
Hey is there any update on this? Thanks

**Comment by toufiqmusah** [comment]
> Hey is there any update on this? Thanks

I've been able to make some 'work arounds' for this. lmk if interested ..

**Comment by joshuacwnewton** [comment]
> Will it be possible to make a pull request that implements this fix in predict_from_raw_data.py?

I think so! I have created a fork of `nnunetv2` with what I believe to be a potential fix: https://github.com/MIC-DKFZ/nnUNet/commit/73ac1f9cf697437eabc1ee33ae0c17d3aec59b3b

The idea is to allow for (optionally) overriding this recursive scan for classes by allowing the user to directly pass a reference to the relevant class instead. That way, the class can live in the user's Python code without the need to modify `nnunetv2` class files.

But, I am still in the process of testing this solution (https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/5003) so it may be a moment before I can open a PR.

**Comment by Karol-G** [comment]
Hey,

Is this issue still relevant? Otherwise, I will close it in 14 days.

Best,
Karol

**Comment by joshuacwnewton** [comment]
Hi @Karol-G! Thanks for asking. Yes, it's still very relevant. This issue is actually the main motivation we had for creating an internal fork of `nnunetv2`. See: Commit https://github.com/spinalcordtoolbox/nnUNet-neuropoly/commit/511204ebed0a780cbdf0e14d011fc037f52f5796 which implements the fix (tl;dr: we want to be able to pass a custom trainer in-memory, rather than putting the custom trainer in the `nnunetv2` package directory).

If this fix was implemented in `nnunetv2`, then we would no longer need to maintain our own fork, and we could go back to using the main version.

**Comment by Karol-G** [comment]
Hi,

nnU-Net is designed as a framework, so changes such as introducing new trainer classes are expected to be implemented within the framework/codebase itself. As @toufiqmusah already explained, the proper way to add and use a new trainer is as follows:

- Create a custom trainer class inside the nnunetv2 codebase by subclassing nnUNetTrainer and overriding the parts you want to change.
- Place that class in the trainer module structure of nnunetv2 so that nnU-Net can discover it, and make sure the class name is unique.
- If you also change the architecture, override build_network_architecture accordingly and ensure the model remains compatible with nnU-Net’s training setup.
- Run training by explicitly specifying your custom trainer in nnUNetv2_train.
- If you also use a custom planner, first run nnUNetv2_plan_and_preprocess with that planner and a custom plans_name, then train with the matching plans and trainer.
- Ensure your trainer properly handles nnU-Net v2 features such as multi-GPU training, region-based training, ignore labels, and cascaded training, or explicitly mark unsupported functionality.

Since you are using nnunetv2 through SCT, you first need to locate the environment in which SCT installed nnU-Net. Then uninstall the nnunetv2 version that comes with SCT and install your modified version instead, ideally in editable mode.

That is the clean and correct way to proceed.

Best,
Karol

**Comment by joshuacwnewton** [comment]
> Since you are using nnunetv2 through SCT, you first need to locate the environment in which SCT installed nnU-Net. Then uninstall the nnunetv2 version that comes with SCT and install your modified version instead, ideally in editable mode.

I understand your points, and the workflow that you describe is tolerable from a research perspective (i.e. students in labs developing custom trainers and running experiments). Our students are OK with placing custom trainers in the `nnunetv2` directory.

> Place that class in the trainer module structure of nnunetv2 so that nnU-Net can discover it, and make sure the class name is unique.

However, this specific step is not feasible from a software development perspective. Specifically, for distributing pretrained models to users outside of our lab, and running inference on user devices. 

- What happens is: We install the "vanilla" `nnunetv2` on a user machine, the user wishes to run inference for a model that used a custom trainer, and now they are responsible for modifying `nnunetv2` on their own (or we have to write some "post-install" code that modifies `nnunetv2` automatically for them). Importantly, however, there is discussion in [many places](https://stackoverflow.com/a/68809248) that `pip` packages should [never be modified after they are installed](https://stackoverflow.com/a/23075617).
- Also, you mention `nnunetv2` being a framework, and I understand. But, by comparison, other deep-learning frameworks (e.g. Tensorflow, Torch, etc.) allow custom architectures to be defined in-code and used in-code. There is no need to put a model architecture in, for example, the `torch` package directory. You simply run the code ([initializing the user-defined model class, loading weights, etc.](https://docs.pytorch.org/tutorials/beginner/saving_loading_models.html#example)) without actually modifying the `torch` directory at all.

The fact that `nnunetv2` requires modifying `./site-packages/nnunetv2` goes against what is standard for `pip` packages. So, I am requesting a way to do "custom trainer" loading in-code, like is common for other frameworks. 

In fact, loading a custom trainer in-code is very easy to do inside of `nnunetv2` with just a small modification -- simply pass the trainer class as an argument, instead of scanning the `nnunetv2` directory, as this commit illustrates: https://github.com/spinalcordtoolbox/nnUNet-neuropoly/commit/511204ebed0a780cbdf0e14d011fc037f52f5796

**Comment by Karol-G** [comment]
Hey,

Thanks for raising this. I think there may be a misunderstanding here.

The intended workflow is **not** to modify the installed package under `./site-packages/nnunetv2`. For development and research with nnU-Net, the recommended approach is to clone the repository, make your changes in that codebase, and then install that version with either `pip install .` or `pip install -e .`.

So if you want to add a custom trainer, you would typically do that in your cloned nnU-Net repository and use that installation for both training and inference, rather than patching files inside an existing site-packages install.

If your goal is to **distribute** a modified version, then the right way is to fork `nnUNet`, implement your changes in that fork, and distribute/install it from there, for example via:

`pip install git+https://github.com/<your-username>/nnUNet.git`

or pinned to a specific branch:

`pip install git+https://github.com/<your-username>/nnUNet.git@<branch-name>`

So in short: no, the expectation is not that users manually edit files inside `site-packages`. For custom development, work from a cloned repo and install that version; for redistribution, use a fork and install from that fork.

Best,
Karol

**Comment by joshuacwnewton** [comment]
> If your goal is to **distribute** a modified version, then the right way is to fork `nnUNet`, implement your changes in that fork, and distribute/install it from there.

I am surprised by this! Forking a package is an extreme action, and imposes quite a lot of maintenance burden on downstream users (syncing upstream changes, mirroring upstream releases, making sure that users properly install the fork instead of accidentally installing the vanilla package, etc.). Forking also encourages rifts within the community, where [development effort is split](https://en.wikipedia.org/wiki/Fork_(software_development)#Forking_of_free_and_open-source_software) when it could be coordinated:

> In a fork, both parties assume nearly identical code bases, but typically only the larger group, or whoever controls the web site, will retain the full original name and the associated user community.
>
> [Eric S. Raymond](https://en.wikipedia.org/wiki/Eric_S._Raymond), in his essay [Homesteading the Noosphere](https://en.wikipedia.org/wiki/Homesteading_the_Noosphere),[[12]](https://en.wikipedia.org/wiki/Fork_(software_development)#cite_note-12) stated that "The most important characteristic of a fork is that it spawns competing projects that cannot later exchange code, splitting the potential developer community".

----

Additionally, the main reason I am surprised is because we don't intend to modify `nnunetv2` itself, i.e. we don't want to distribute a version of `nnunetv2` that contains the custom trainer files. We simply want the ability for custom trainers to be passed to `nnunetv2` for inference (https://github.com/spinalcordtoolbox/nnUNet-neuropoly/commit/511204ebed0a780cbdf0e14d011fc037f52f5796). The linked commit is tiny, so forking the full `nnunetv2` package feels unnecessary / disproportionate to the changes involved.

> _(In fact, we are already working with a version of `nnunetv2` that contains the above commit, which allows us to distribute our custom trainer files [alongside versioned model checkpoints](https://intranet.neuro.polymtl.ca/data/deeplearning-models.html#packaging-models), separate from the `nnunetv2` package itself. This ties the custom trainer to the model release, rather than the release of `nnunetv2`, which we decided on for better provenance of the model source.)_

If forking is really the official stance, then we will continue with the above approach. But, we would strongly prefer to avoid forking and instead continue to use the mainline package (so that we can contibute upstream bugfixes/improvements/etc. and stay in sync with the main package more easily).

**Comment by Karol-G** [comment]
Hey,

You asked for clarification on the currently recommended way to implement and use a new nnU-Net trainer, and I responded with what is currently the recommended workflow. There is no need for the attitude.

That being said, if your point is that nnU-Net should support custom trainers more generally, without requiring users to distribute a modified nnU-Net version, then I do understand that request. I can forward that request to Fabian, the main nnU-Net author.

However, I think it would be better to open a separate issue specifically for that feature request. In that issue, it would help to describe the use case more clearly and explain the broader benefit of such functionality. Fabian usually prefers to implement or accept features when there is a clear, strong, and general benefit.

If you would like to open such a dedicated issue, I would be happy to forward it to him. Otherwise, I can also forward the current issue as it stands.

Best,
Karol

**Comment by joshuacwnewton** [comment]
> There is no need for the attitude.

My apologies -- I'm sorry for bringing negativity to the conversation. Tone is hard over text and I spoke using terse/blunt language.

> However, I think it would be better to open a separate issue specifically for that feature request. In that issue, it would help to describe the use case more clearly and explain the broader benefit of such functionality.

Sure, I can do that (https://github.com/MIC-DKFZ/nnUNet/issues/3025). :)

**Comment by joshuacwnewton** [comment]
Closing this issue as resolved, as the question has been answered. 

tl;dr summary of the current recommendations:

- **Training/development:** Clone `nnunetv2`, install in editable mode, place the custom trainer in the `nnunetv2` directory.
- **Distribution:** Fork `nnunetv2`, place the custom trainer in the `nnunetv2` directory, install by pointing `pip` at the GitHub repo.
